linalg/batched/batched.hpp

```diff
@@ -0,0 +1,130 @@
+// Copyright (c) 2010-2024, Lawrence Livermore National Security, LLC. Produced
+// at the Lawrence Livermore National Laboratory. All Rights reserved. See files
+// LICENSE and NOTICE for details. LLNL-CODE-806117.
+//
+// This file is part of the MFEM library. For more information and source code
+// availability visit https://mfem.org.
+//
+// MFEM is free software; you can redistribute it and/or modify it under the
+// terms of the BSD-3 license. We welcome feedback and contributions, see file
+// CONTRIBUTING.md for details.
+
+#ifndef MFEM_BATCHED_LINALG
+#define MFEM_BATCHED_LINALG
+
+#include "../../config/config.hpp"
+#include "../densemat.hpp"
+#include <array>
+#include <memory>
+
+namespace mfem
+{
+
+/// @brief Class for performing batched linear algebra operations, potentially
+/// using accelerated algorithms (GPU BLAS or MAGMA). Accessed using static
+/// member functions.
+///
+/// The static member functions will delegate to the preferred backend (which
+/// can be set using SetPreferredBackend(), see BatchedLinAlg::Backend for all
+/// available backends and the order in which they will be chosen by default).
+/// Operations can be performed directly with a specific backend using Get().
+class BatchedLinAlg
+{
+public:
+   /// @brief Available backends for implementations of batched algorithms.
+   ///
+   /// The preferred backend will be the first available backend in this order:
+   /// MAGMA, GPU_BLAS, NATIVE.
+   enum Backend
+   {
+      /// @brief The standard MFEM backend, implemented using mfem::forall
+      /// kernels. Not as performant as the other kernels.
+      NATIVE,
+      /// @brief Either cuBLAS or hipBLAS, depending on whether MFEM is using
+      /// CUDA or HIP. Not available otherwise.
+      GPU_BLAS,
+      /// MAGMA backend, only available if MFEM is compiled with MAGMA support.
+      MAGMA,
+      /// Counter for the number of backends.
+      NUM_BACKENDS
+   };
+private:
+   /// All available backends. Unavailble backends will be nullptr.
+   std::array<std::unique_ptr<class BatchedLinAlgBase>,
+          Backend::NUM_BACKENDS> backends;
+   Backend preferred_backend;
```

Comment:
Similarly here, using "preferred_*" kind of implies that this may not be the actual backend that will be used. I'd just call this "backend".

Reply:
Changed in 3a5928160192cdd8ab995ef46cbd1181f4fb0318.